### DISCUSSION TITLE
Weird resize artifacts

### DISCUSSION THREAD

**Discussion by exekutive** [ImageMagick/ImageMagick#7404] [Help] [answered]
### ImageMagick version

7.1.1-31 (Beta) Q16-HDRI

### Operating system, version and so on

darwin aarch64

### Description

Taken from Robidoux's "Recommended Upsampling Methods"
(https://imagemagick.org/Usage/filter/nicolas/#upsampling)

```
magick $f -colorspace RGB +sigmoidal-contrast 7.5 \
          -filter LanczosRadius -distort Resize 500% \
          -sigmoidal-contrast 7.5 -colorspace sRGB ${f:r}RR2.png
```

Source:
[header_.png.zip](https://github.com/user-attachments/files/15862400/header_.png.zip)

![header_](https://github.com/ImageMagick/ImageMagick/assets/4755215/af2e8969-bded-42c5-8d23-3226a4421d36)

Distort resized:
![header_RR2](https://github.com/ImageMagick/ImageMagick/assets/4755215/1ff82997-daa8-41d6-a293-a2e7790c81db)

closeup of speckling:

![speckle](https://github.com/ImageMagick/ImageMagick/assets/4755215/090670b4-0b74-43a9-bc32-c963bdd16069)

**Comment by exekutive** [comment]
artifact does not appear if resize is set to 400% instead of 500%

**Reply by fmw42** [reply]
I see the same artifacts in your input image as in my output image when using IM 7.1.1.33 Mac OSX Ventura. The artifacts from the input are magnified in the output as expected.

![header_](https://github.com/ImageMagick/ImageMagick/assets/13458218/3237eb57-7fc9-4b89-9abd-0065133ea80e)


```
magick image.png -colorspace RGB +sigmoidal-contrast 7.5 \
-filter LanczosRadius -distort Resize 500% \
-sigmoidal-contrast 7.5 -colorspace sRGB x.png
```

![x](https://github.com/ImageMagick/ImageMagick/assets/13458218/580f9415-ac31-4cda-b5e9-11019845f527)

**Reply by exekutive** [reply]
@fmw42 
you are looking at the wrong image. Sorry, maybe I should have labeled them better.
input image is 219 x 66. Check the zip file.

**Reply by fmw42** [reply]
There is only one image in the zip file which is what I used above.  At first I had just grabbed the image from above, but then I saw your zip file and corrected it.  In either case, the flaws are in your input image as best I can see, if I magnify your input and compare to the magnified output.

**Reply by exekutive** [reply]
No. You are misunderstanding or doing something wrong. This is how the source image looks up close. No speckles.

![h](https://github.com/ImageMagick/ImageMagick/assets/4755215/cb9cfb7a-797a-4d5e-b00b-413fcd498bf9)

**Reply by exekutive** [reply]
and here is the result if I use 400% instead of 500%. No speckles.

```
magick image.png -colorspace RGB +sigmoidal-contrast 7.5 \
-filter LanczosRadius -distort Resize 400% \
-sigmoidal-contrast 7.5 -colorspace sRGB x.png
```
![h2](https://github.com/ImageMagick/ImageMagick/assets/4755215/63ffb3ba-83fc-4b5a-b8cc-3222340de6d8)

**Comment by snibgo** [comment]
The image in the zip file header_.png.zip doesn't have speckles. The result after `-filter LanczosRadius -distort Resize 500% ` does have speckles.

We can reproduce the problem with a different input. Windows BAT syntax:
```
magick ^
  -fill Red -background Green ^
  caption:Hello ^
  -trim +repage ^
  +write c.png ^
  -filter LanczosRadius -distort Resize 500%% -clamp ^
  -scale 1000%% ^
  c2.png
```
The input to the resize, `c.png`, is:

![c](https://github.com/ImageMagick/ImageMagick/assets/33812368/bf233d33-b90f-400d-9694-f75843c80031)

The output of the resize, scaled up by 1000% to make the problem obvious, is:

![c2](https://github.com/ImageMagick/ImageMagick/assets/33812368/d65a882e-c56d-42f8-8322-feed82a47f5f)

The speckles occur at an interval of 5 pixels in each direction. It also occurs when resizing to 300% or 700%, at intervals of 3 or 7 pixels respectively.

Is this a bug, or a "feature" of LanczosRadius? I don't know.

**Comment by fmw42** [comment]
Yes, I saw artifacts only in the first input image that downloaded from his first post before I saw the zip link.  Sorry, I had just grabbed the image that he posted thinking it was his original. I did not see artifacts in the smaller header_.png image.

**Comment by exekutive** [comment]
Thank you for confirming @snibgo.

Maybe @nrobidoux can shed some light on this.

**Comment by nrobidoux** [comment]
Are you using a floating point or integer pipeline? (Been "just a user" for a long time now so I don't know if the floating point/integer versions still exist.) These look like integer overflow artifacts. From memory: I remember very carefully putting together sigmoidal contrast so it would correctly "overshoot" and "undershoot" but maybe I was not so careful for the integer pipelines. I'll have to remind myself of the finer points of IM pipelines but can you "force" a floating point pipeline through some conversion? (Sorry, I'm really busy today so will take some time.)

**Comment by snibgo** [comment]
My tests were with floating point. IM v7.1.1-20 on Windows 11.

IM can be built so pixel values are stored in memory either as floating-point or integer. `magick -version`, under `Features:`, if that includes `HDRI`, IM is built with floating-point. Otherwise it is integer.

**Comment by nrobidoux** [comment]
Snibgo: Sorry, catching up with the reading. I see your test pipeline does not contain signmoidal-contrast so it's not that.

**Comment by nrobidoux** [comment]
I persist in thinking that this is an integer overflow/underflow issue. But I'll have to dig. It >>is<< possible that LanczosRadius (which from a theoretical viewpoint is not as sound a filter than others) produces negative sums of weights in some circumstances in which case it would be a "theoretical" bug (my fault) not a "code" bug (possibly not my fault) but for me to check that I'd need to invest time. Enlarging by an exact odd integer may be exactly the kind of circumstance what reveals it, and then LanczosRadius should NOT be recommended for enlarging. Maybe early next week. Note: My IM recommendations really date now.

**Comment by nrobidoux** [comment]
Me debugging this (whether theoretical or code bug) should not be a huge deal but I need to get myself back up to speed.

**Comment by nrobidoux** [comment]
One thing that would help me is if someone could check if this happens with old versions of IM.

**Comment by nrobidoux** [comment]
Ty exekutive for bringing this up to our attention.

**Comment by snibgo** [comment]
With IM 6.9.12-72, with HDRI (hence recording pixels as floating-point):
```
convert c.png -filter LanczosRadius -distort Resize 500%% -scale 1000%% x.png
```
x.png is:

![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/456508a2-24f6-4731-81d4-4dc628139d0e)

Thus, the problem does not occur.

EDIT: Correction, this is with HDRI.

**Comment by nrobidoux** [comment]
Ty Snibgo. I note you removed the -clamp.

**Comment by nrobidoux** [comment]
Pain in the neck of course but at this point I'm inclined to think that most efficient way to figure what's going on is to bisect until we know what code change led to the issue.

**Comment by nrobidoux** [comment]
Anyway these artifacts look to me like "wrap around" hence incorrect cast to a "small" unsigned integer type, in which case I would expect such artifacts to occur with other filters ("anything" with negative lobes).

**Comment by snibgo** [comment]
I removed `-clamp` as it makes no difference to the result. 

Lanczos2 has a similar problem to LanczosRadius:

![rt_Lanczos2](https://github.com/ImageMagick/ImageMagick/assets/33812368/ccded73e-56f2-4ee4-8418-4206c00754f1)

Trying all the filters, I notice problems with Sinc and SincFast. The oddities there are at the "centres" of the speckles from LanczosRadius. 

Sinc filter:

![rt_Sinc](https://github.com/ImageMagick/ImageMagick/assets/33812368/66c99569-ec90-47b5-a2d6-08672d30b5d7)

SincFast filter:

![rt_SincFast](https://github.com/ImageMagick/ImageMagick/assets/33812368/1267673f-c93e-4c7e-b9c3-863883644526)

**Comment by nrobidoux** [comment]
This breaks enough stuff that the bug is most likely to have been introduced recently.

**Comment by nrobidoux** [comment]
When you enlarge with a filter which has support exactly 3 (like Sinc, SincFast, LanczosRadius, all with default 3) when you are close to an original pixel you use 5 pixels in both directions and the largest weight after normalization is larger than 1. Probably true with Lanczos as well (which does NOT have support exactly 3, but close enough).

**Comment by nrobidoux** [comment]
Anyway, when I look, I'll know what to look for: handling of intermediate results larger than "max" or negative.

**Comment by nrobidoux** [comment]
I honestly don't think it's something I did (or didn't do). Which does not mean I can't help.

**Comment by exekutive** [comment]
Looking at the resized image in my original post, the speckles are not evenly distributed across the whole image. They are concentrated in certain areas. Maybe that is a clue.
Is there some way to output intermediate values to help narrow down where in the pipeline the errant pixels are arising?
Surely this is unexpected behaviour, right? Filters should not just garble our images. All the arguments seem fairly safe. Should this be a bug report?

**Reply by nrobidoux** [reply]
It is. What it says is that the garbage arises from pixels near the edge of the filter's support. This "edge" slowly drifts relative to pixel positions, a bit like "moire".

**Comment by exekutive** [comment]
I can confirm it has nothing to do with sigmoidal contrast or colorspace.
`-clamp` , `-depth 8` and 16, and `-alpha off` have no effect

This simple command still produces speckles.

`magick header_.png -filter LanczosRadius -distort Resize 500% speckle.png`

**Comment by nrobidoux** [comment]
Yes it should be a bug report.

**Comment by fmw42** [comment]
I still see these artifacts in the output under 6.9.13-11 Q16 and 6.9.12.80 Q16. (non-hdri).

```
convert header_.png -colorspace RGB +sigmoidal-contrast 7.5 \
-filter LanczosRadius -distort Resize 500%  \
-sigmoidal-contrast 7.5 -colorspace sRGB x500.png

```

I also tried -clamp in a number of places under IM 7.1.1.33 Q16 HDRI, but the artifacts are still there.  I even tried Q32 HDRI and it still had the artifacts.  I also went back to 7.1.1.15 and it still had the artifacts

**Reply by nrobidoux** [reply]
So it probably happened after 6.9.12-72 and before 6.9.12-80.

**Comment by exekutive** [comment, marked as answer]
submitted issue https://github.com/ImageMagick/ImageMagick/issues/7415
